src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -269,15 +269,21 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for UsedGenericParametersVisitor<'a, 'tcx> {
                 self.unused_parameters.clear(param.index);
                 false
             }
-            ty::ConstKind::Unevaluated(_, _, Some(p)) => {
+            ty::ConstKind::Unevaluated(def, _, Some(p))
+                // Avoid considering `T` unused when constants are of the form:
+                //   `<Self as Foo<T>>::foo::promoted[p]`
+                if !self.tcx.generics_of(def.did).has_self =>
+            {
                 // If there is a promoted, don't look at the substs - since it will always contain
                 // the generic parameters, instead, traverse the promoted MIR.
-                let promoted = self.tcx.promoted_mir(self.def_id);
+                let promoted = self.tcx.promoted_mir(def.did);
                 self.visit_body(&promoted[p]);
```

Comment:
What happens here if the `def.did != self.def_id`. Aren't we then treating the `ty::Param` of the promoted mir as params of `self.def_id`?

Should we just constrain this to `if !self.tcx.generics_of(def.did).has_self && self.def_id == def.did` for now or am I misunderstanding something here?

Reply:
> What happens here if `def.did != self.def_id`. Aren't we then treating the `ty::Param` of the promoted mir as params of `self.def_id`?

As far as I'm aware, `def.did` should be the same as `self.def_id` (unless we're in the MIR inliner, apparently - which shouldn't be a problem for us). The assumption here is that the promoted MIR has the same generic params as the parent fn's MIR - which is the same assumption that the original PR adding this support had.

Reply:
> unless we're in the MIR inliner, apparently - which shouldn't be a problem for us

not unless we're *in* the MIR inliner, unless we *use* the mir inliner. The following breaks:
```rust
// compile-flags: -Zpolymorphize=on -Zmir-opt-level=3
#![feature(rustc_attrs)]

fn caller<T, U>() -> &'static usize {
    callee::<U>()
}

fn callee<T>() -> &'static usize {
    &std::mem::size_of::<T>()
}

fn main() {
    assert_eq!(caller::<(), ()>(), &7);
}
```
> thread 'rustc' panicked at 'index out of bounds: the len is 0 but the index is 0', src/librustc_mir/monomorphize/polymorphize.rs:276:34

Reply:
https://rust.godbolt.org/z/s8rd99

Reply:
Fixed!